### PR TITLE
enhancement(sampler transform): Dynamic sampling RFC

### PR DESCRIPTION
**Pull request by lucperkins**
This RFC proposes sweeping changes to the `sampler` transform that would make it support dynamic sampling and thus cover a much wider range of sampling use cases. It's a work in progress but I'm opening it up for an initial comment/critique window.

### CODE REVIEW

File: rfcs/2020-12-08-5448-dynamic-sampling.md [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+  plus some additional explanation.
+* A new dynamic sampling guide that walks through specific sampling use cases.
+* A new "Under the Hood" page à la the one we provide for Adaptive Request Concurrency that explains
+  the feature in more detail, with accompanying diagrams
+
+## Prior Art
+
+Other systems in the observability space do provide dynamic sampling capabilities. Their feature
+sets and user interfaces can provide inspiration for dynamic sampling in Vector (and have influenced
+this RFC). In terms of implementation, however, I'm unaware of a system built in Rust that provides
```
Comment (jszwedko) [inline rfcs/2020-12-08-5448-dynamic-sampling.md:325]: We've often linked to prior art of other applications (like fluentbit, logstash, etc.) regardless of their implementation language. It is still useful to compare their implementation and the configuration options they chose to expose.

Would you be able to add some you are aware of?

I did find something kinda analogous in logstash: https://www.elastic.co/guide/en/logstash/current/plugins-filters-drop.html

**Review by jszwedko — COMMENTED** [review]
This is cool! Thanks @lucperkins . The example is motivating.

I would like to see the example you have here using swimlanes just for comparison.

I'm also wondering if we have any intentions to support sampling in `remap` cc/ @binarylogic @JeanMertz @FungusHumungus as that could affect this.

I'm a little confused about the new sampling parameters (`max_rate`/`min_rate`, `sensitivity`, `min_event_threshold`, `window_secs`). Would you be able to explain a bit where those would be more useful than the existing `rate` parameter? An example might help too. It also seems like that bit could be split from this RFC if needed, to just focus on enhancing the `sampler` transform to allow differing the rates based on conditions.

I actually got the feeling, reading this, that this sort of pattern could be generalized to any transform. Like `remap` as well:

```toml
[transforms.my_transform_id]
  type = "remap" # required

[[bucket.less_interesting_transactions]]
condition = """
.status == 200 && !in_list(.user.plan, ["pro", "enterprise", "deluxe"]) && .transaction.accepted == true
"""
source = """ do something """

[[bucket.very_interesting_because_catastrophic_transactions]]
condition = """
.status == 500 && .user.plan == "deluxe" && len(.shopping_cart.items) > 10
"""
source = """ do something else"
```

File: rfcs/2020-12-08-5448-dynamic-sampling.md [hunk 2]
```diff
@@ -0,0 +1,351 @@
+# RFC 5448 - <2020-12-08> - Dynamic sampling in the `sampler` transform
```
Comment (jszwedko) [inline rfcs/2020-12-08-5448-dynamic-sampling.md:1]: Just noting this RFC issue number is pointing to #5448 which doesn't seem right. I think you probably need to open a separate RFC issue like #2414 and then link that here

File: rfcs/2020-12-08-5448-dynamic-sampling.md [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+Named buckets are useful because they enable you to route events using arbitrarily complex criteria.
+Any Remap expression that returns a Boolean can be used as a condition.
+
+#### Order of application
+
+The `sampler` transform should place events in buckets in the order in which the conditions are
+specified, much like a `switch` statement in many programming languages. An event ends up
+in the first bucket whose condition it matches; when a match occurs, condition checking ceases.
+
+This raises the question of fallthrough behavior, i.e. what happens if an event doesn't meet any
```
[the remark this answers is not in the document]
Comment (lucperkins) [reply rfcs/2020-12-08-5448-dynamic-sampling.md]: Good to know! I'll update with a reference to that issue.

**Comment by lucperkins** [comment]
@jszwedko I think you're right that the actual dynamic sampling part gets a bit buried by the bucketing part here. So I'm going to trim this down by removing the "named buckets" content and deferring to other initiatives that are already under way, because it sounds like that part of the RFC may be less than helpful as part of a chorus of related discussions.

File: rfcs/2020-12-08-5457-dynamic-sampling.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+### Domains
+
+Dynamic sampling has implications for both logging and tracing. It does *not* have implications for
+metrics, as metrics are already a form of sampling in themselves, e.g. counters reveal a raw number
+of events while saying little about the content of those events.
+
+This RFC, however, only covers logging and not tracing. It's possible that tracing would be
+seamlessly covered by the changes proposed here but I lack the domain-specific knowledge to assert
+that, so I will leave it as an open question for a future RFC.
```
Comment (binarylogic) [inline rfcs/2020-12-08-5457-dynamic-sampling.md:33]: Agree. Sampling traces is much much more complicated. Look at the Lightstep agent for prior art.

**Comment by binarylogic** [comment]
Thanks for putting this together @lucperkins. I think we should hold off on discussing this until we are ready to work on it. This is an involved topic that we'll want to research, as a team, before proceeding. We currently do not have the bandwidth across the team to properly move forward on this. We should revive this RFC when the time is right, which should be in the near term.

**Pull request closed without merging**